README.rst
.. contents::

For usage information, please see the tests at bbe/cielo/tests.py


CHANGES.txt
Changelog
=========

0.0.1-dev (unreleased)
----------------------

- Package created using templer
  []

bbe/cielo/tests.py
# -*- coding: utf-8 -*-
from decimal import Decimal
import colander
import datetime
import unittest
import bbe.cielo as cielo


def nextmonth():
    return datetime.datetime.now() + datetime.timedelta(days=30)


class MessageSerializationTestCase(unittest.TestCase):
    def assertDumps(self, node, appstruct, test):
        cstruct = node.serialize(appstruct)
        etree = cielo.message.serialize(node, cstruct)
        s = cielo.message.dumps(etree)
        return self.assertEqual(s, test)

    def test_serialization(self):
        node = colander.SchemaNode(colander.String(), name='node')
        self.assertDumps(node, 'abcdef', '<node>abcdef</node>')

    def test_serialization_with_tag(self):
        node = colander.SchemaNode(colander.String(), name='node', tag='node-tag')
        self.assertDumps(node, 'abcdef', '<node-tag>abcdef</node-tag>')

    def test_subnode_serialization(self):
        node = colander.SchemaNode(colander.Mapping(), name='node')
        node.add(colander.SchemaNode(colander.String(), name='sub'))

        self.assertDumps(node, {'sub': 'abcdef'}, '<node><sub>abcdef</sub></node>')

    def test_subnode_serialization_order(self):
        "Subnode serialization must follow schema's order"
        node = colander.SchemaNode(colander.Mapping(), name='node')
        node.add(colander.SchemaNode(colander.String(), name='a'))
        node.add(colander.SchemaNode(colander.String(), name='b'))
        node.add(colander.SchemaNode(colander.String(), name='c'))

        appstruct = {'a': 'value-1', 'b': 'value-2', 'c': 'value-3'}

        #self.assertNotEqual(appstruct.keys(), ['a', 'b', 'c'])
        self.assertDumps(node, appstruct, '<node><a>value-1</a><b>value-2</b><c>value-3</c></node>')

    def test_empty_mapping_serialization(self):
        node = colander.SchemaNode(colander.Mapping(), name='node')
        node.add(colander.SchemaNode(colander.String(), name='a'))
        node.add(colander.SchemaNode(colander.String(), name='b'))
        node.add(colander.SchemaNode(colander.String(), name='c'))
        self.assertDumps(node, {}, '<node />')

    def test_null_mapping_serialization(self):
        node = colander.SchemaNode(colander.Mapping(), name='node')
        node.add(colander.SchemaNode(colander.String(), name='a'))
        node.add(colander.SchemaNode(colander.String(), name='b'))
        node.add(colander.SchemaNode(colander.String(), name='c'))
        self.assertDumps(node, colander.null, '<node />')


class MessageDeserializationTestCase(unittest.TestCase):
    def assertLoads(self, node, message, test):
        etree = cielo.message.loads(message)
        cstruct = cielo.message.deserialize(node, etree)
        appstruct = node.deserialize(cstruct)
        return self.assertEqual(appstruct, test)

    def test_serialization(self):
        node = colander.SchemaNode(colander.String(), name='node')
        self.assertLoads(node, '<node>abcdef</node>', 'abcdef')

    def test_deserialization_with_tags(self):
        node = colander.SchemaNode(colander.String(), name='node', tag='node-tag')
        self.assertLoads(node, '<node-tag>abcdef</node-tag>', 'abcdef')

    def test_subnode_deserialization(self):
        node = colander.SchemaNode(colander.Mapping(), name='node')
        node.add(colander.SchemaNode(colander.String(), name='sub'))
        self.assertLoads(node, '<node><sub>abcdef</sub></node>', {'sub': 'abcdef'})

    def test_empty_mapping_deserialization(self):
        node = colander.SchemaNode(colander.Mapping(), name='node')
        node.add(colander.SchemaNode(colander.String(), name='a', missing=colander.null))
        node.add(colander.SchemaNode(colander.String(), name='b', missing=colander.null))
        node.add(colander.SchemaNode(colander.String(), name='c', missing=colander.null))
        self.assertLoads(node, '<node />', {'a': colander.null, 'b': colander.null, 'c': colander.null})

class MoneyTestCase(unittest.TestCase):
    def setUp(self):
        self.node = colander.SchemaNode(cielo.Money())

    def assertSerialize(self, appstruct, cstruct):
        self.assertEqual(self.node.serialize(appstruct), cstruct)

    def assertDeserialize(self, cstruct, appstruct):
        self.assertEqual(self.node.deserialize(cstruct), appstruct)

    def test_null_serialization(self):
        self.assertSerialize(colander.null, colander.null)

    def test_null_deserialization(self):
        node = self.node.clone()
        node.missing = colander.null
        self.assertEqual(node.deserialize(colander.null), colander.null)

    def test_integer_serialization(self):
        self.assertSerialize(188, '18800')

    def test_float_serialization(self):
        self.assertSerialize(333.0, '33300')
        self.assertSerialize(333.1, '33310')
        self.assertSerialize(333.11, '33311')

    def test_decimal_serialization(self):
        self.assertSerialize(Decimal('123.00'), '12300')
        self.assertSerialize(Decimal('123.5'), '12350')
        self.assertSerialize(Decimal('123.56'), '12356')

    def test_deserialization(self):
        self.assertDeserialize('43200', Decimal('432.00'))
        self.assertDeserialize('87720', Decimal('877.2'))
        self.assertDeserialize('12311', Decimal('123.11'))

    def test_non_numeric_values_serialization(self):
        "Obviously, non-numeric vaulues are invalid"
        self.assertRaises(colander.Invalid, self.node.serialize, None)
        self.assertRaises(colander.Invalid, self.node.serialize, 'notanumber')

    @unittest.skip("i'm not sure about this one")
    def test_string_numeric_values_deserialization(self):
        self.assertRaises(colander.Invalid, self.node.serialize, '42')

    def test_non_numeric_values_deserialization(self):
        "Obviously, non-numeric values will be invalid during deserialization"
        self.assertRaises(colander.Invalid, self.node.deserialize, None)
        self.assertRaises(colander.Invalid, self.node.deserialize, 'notanumber')

    def test_invalid_monetary_values(self):
        "Values with more than two decimal places are invalid"
        self.assertRaises(colander.Invalid, self.node.serialize, 100.123)
        self.assertRaises(colander.Invalid, self.node.serialize, Decimal('200.543'))


# do not trust these

class TestCase(unittest.TestCase):
    def setUp(self):
        self.client = cielo.Client(
            #store_id='1001734898',
            #store_key='e84827130b9837473681c2787007da5914d6359947015a5cdb2b8843db0fa832',
            store_id='1006993069',
            store_key='25fbb99741c739dd84d7b06ec78c9bac718838630f30b112d033ce2e621b34f3',
            service_url='https://qasecommerce.cielo.com.br/servicos/ecommwsec.do',
        )


class MonolithicTestCase(TestCase):
    def test_process_response(self):
        response = self.client.process_response(u"""
            <transacao versao="1.1.1" id="f71e286f-21f6-4abe-8999-cc200e585454" xmlns="http://ecommerce.cbmp.com.br">
              <tid>100699306905227C1001</tid>
              <pan>uv9yI5tkhX9jpuCt+dfrtoSVM4U3gIjvrcwMBfZcadE=</pan>
              <dados-pedido>
                <numero>1</numero>
                <valor>20000</valor>
                <moeda>96</moeda>
                <data-hora>2012-08-11T08:48:23.659-03:00</data-hora>
                <idioma>PT</idioma>
              </dados-pedido>
              <forma-pagamento>
                <bandeira>visa</bandeira>
                <produto>1</produto>
                <parcelas>1</parcelas>
              </forma-pagamento>
              <status>5</status>
              <autenticacao>
                <codigo>5</codigo>
                <mensagem>Transacao sem autenticacao</mensagem>
                <data-hora>2012-08-11T08:48:23.695-03:00</data-hora>
                <valor>20000</valor>
                <eci>7</eci>
              </autenticacao>
              <autorizacao>
                <codigo>5</codigo>
                <mensagem>Autorização negada</mensagem>
                <data-hora>2012-08-11T08:48:43.708-03:00</data-hora>
                <valor>20000</valor>
                <lr>999</lr>
                <nsu>336508</nsu>
              </autorizacao>
            </transacao>""")
        self.assertTrue(response.success)
        self.assertIsInstance(response, cielo.Transaction)
        self.assertEqual(response.status, 5)

    def test_process_error_response(self):
        response = self.client.process_response(
            u"""<?xml version="1.0" encoding="ISO-8859-1"?>
                <erro>
                    <codigo>032</codigo>
                    <mensagem>Valor de captura inválido</mensagem>
                </erro>""")
        self.assertFalse(response.success)
        self.assertEqual(response.code, 32)
        self.assertEqual(response.message, u"Valor de captura inválido")

    @unittest.skip("this is not finished yet")
    def test_transaction(self):
        order = cielo.Order(value=Decimal('200.0'))
        payment = cielo.Payment(
            card_number='4551870000000183',
            card_expiration_date=nextmonth(),
            card_security_code='123',
            card_holder_name='Joao da Silva',
            card_brand=cielo.VISA,
            value=order.value,
            datetime=order.datetime,
            installments=1,
        )
        self.client.create_transaction(
            order=order,
            payment=payment,
            authorize=3,
            capture=True,
            return_url='http://example.com',
        )

"""
As informações abaixo podem ser usadas pelo desenvolvedor durante o desenvolvimento da
integração.
Cartão com autenticação: 4012001037141112 (visa)
Cartão sem autenticação: 4551870000000183 (visa), 5453010000066167 (mastercard),
6362970000457013 (elo), 36490102462661 (diners) e 6011020000245045 (discover).
Data de validade: qualquer posterior ao corrente
Código de segurança: qualquer
Valor do pedido: para simular transação autorizada, use qualquer valor em que os dois
últimos dígitos sejam zeros. Do contrário, toda autorização será negada.
Credenciais
Leitura
Número
 Chave
do
cartão
Loja
 1006993069 25fbb99741c739dd84d7b06ec78c9bac718838630f30b112d033ce2e621b34f3
Cielo
 1001734898 e84827130b9837473681c2787007da5914d6359947015a5cdb2b8843db0fa832

"""


bbe/cielo/client.py
# -*- coding: utf-8 -*-
import uuid
import urllib2
import contextlib
from bbe.cielo import message
from bbe.cielo.schema import (
    TransactionRequestSchema,
    INFORMADO,
    INEXISTENTE,
    DEBITO,
    CREDITO_A_VISTA,
    PARCELADO_ADMINISTRADORA,
    guess_response_schema,
    DebitPayment,
)


class CommunicationError(urllib2.URLError):
    """This exception is raised when the communication between
    our client and the remote service fail.

    .. attribute:: reason

        The error reason. It can be a message or an instance of
        another exception.
    """


class UnknownResponse(Exception):
    """This is raised when our client don't know how to handle
    the response.
    """


class Client(object):
    def __init__(self, store_id, store_key, service_url=None):
        self.store_id = store_id
        self.store_key = store_key
        self.service_url = service_url

    def generate_transaction_id(self):
        return str(uuid.uuid4())

    def create_transaction(self, order, payment,  authorize, capture, return_url=None):
        if payment.card_security_code is None:
            code_indic = INEXISTENTE
        else:
            code_indic = INFORMADO
            # TODO: support other security_code indices

        if isinstance(payment, DebitPayment):
            product = DEBITO
        elif payment.installments == 1:
            product = CREDITO_A_VISTA
        else:
            product = PARCELADO_ADMINISTRADORA
            # TODO suportar parcelado_administradora

        appstruct = {
            'id': self.generate_transaction_id(),
            'version': '1.1.1',
            'establishment': {
                'number': self.store_id,
                'key': self.store_key,
            },
            'order': {
                'number': order.number,
                'value': order.value,
                'currency': order.currency,
                'datetime': order.datetime,
                'description': order.description,
                'language': order.language,
            },
            'holder': {
                'number': payment.card_number,
                'security_code': payment.card_security_code,
                'security_code_indicator': code_indic,
                'expiration_date': payment.card_expiration_date,
                'holder_name': payment.card_holder_name,
            },
            'payment': {
                'card_brand': payment.card_brand,
                'product': product,
                'installments': payment.installments,
            },
            'return_url': return_url,
            'authorize': authorize,
            'capture': capture,
            'bin': payment.card_number[:6],
        }

        schema = TransactionRequestSchema(tag='requisicao-transacao')
        request = schema.serialize(appstruct)
        request = message.dumps(request)
        return self.post_request(request)

    def post_request(self, request):
        msg = u"mensagem=" + request
        #TODO msg = msg.encode('iso-')

        try:
            request = urllib2.urlopen(self.service_url, msg)
            with contextlib.closing(request) as response:
                response = response.read()
        except urllib2.URLError, e:
            raise CommunicationError(e.reason)

        return self.process_response(response.decode('iso-8859-1'))

    def process_response(self, response):
        response = message.loads(response.encode('utf-8'))

        # get the root type
        root_tag = message.get_root_tag(response)

        # let's try to guess the response content-type
        schema = guess_response_schema(root_tag)

        if schema is None:
            raise UnknownResponse("Unknown response type '%s'" % root_tag)

        # TODO specialized exception class?
        cstruct = message.deserialize(schema, response)
        response = schema.deserialize(cstruct)
        response = schema.content_type.deserialize(response)

        return response


bbe/cielo/message.py
import re
import colander
import contextlib
from xml.etree.ElementTree import ElementTree, Element, fromstring

try:
    from cStringIO import StringIO
except:
    from StringIO import StringIO


def isattrib(node):
    return getattr(node, 'attrib', False)


def gettag(node):
    return getattr(node, 'tag', node.name)


def _build_element(node):
    return Element(gettag(node))


def serialize(schema, cstruct):
    if cstruct is colander.null:
        return None

    if isinstance(schema.typ, colander.Mapping):
        return _serialize_mapping(schema, cstruct)
    else:
        element = _build_element(schema)
        element.text = cstruct
        return element


def _serialize_mapping(schema, cstruct):
    element = _build_element(schema)

    for child in schema:
        subtag = gettag(child)
        subvalue = cstruct.get(child.name, colander.null)

        # TODO: default, required, etc
        if subvalue is colander.null:
            continue

        if isattrib(child):
            element.attrib[subtag] = subvalue
        else:
            subelement = serialize(child, subvalue)
            if subelement is not None:
                element.append(subelement)

    return element


def deserialize(schema, etree):
    if isinstance(schema.typ, colander.Mapping):
        return _deserialize_mapping(schema, etree)

    if etree.text is None:
        return colander.null

    return etree.text


def _deserialize_mapping(schema, etree):
    cstruct = {}
    for child in schema.children:
        tag = gettag(child)
        if isattrib(child):
            value = etree.attrib.get(tag, colander.null)
        else:
            subelement = etree.find(tag)
            if subelement is None:
                value = colander.null
            else:
                value = deserialize(child, subelement)
        cstruct[child.name] = value
    return cstruct


def dumps(etree):
    if not isinstance(etree, ElementTree):
        etree = ElementTree(etree)
    with contextlib.closing(StringIO()) as s:
        etree.write(s)
        return s.getvalue()


def loads(string):
    etree = fromstring(string)
    remove_namespaces(etree)
    return etree


def remove_namespaces(element):
    """Remove all namespaces in the passed element in place."""
    for ele in element.getiterator():
        ele.tag = re.sub(r'^\{[^\}]+\}', '', ele.tag)


def get_root_tag(etree):
    return etree.tag


bbe/cielo/schema.py
# -*- coding: utf-8 -*-
import inspect
import datetime
import colander
from decimal import Decimal


SERVICE_VERSION = '1.1.1'


LANG_PT = 'PT'
LANG_EN = 'EN'
LANG_ES = 'ES'

LANGUAGES = (LANG_PT, LANG_EN, LANG_ES)


MASTERCARD = 'mastercard'
DINERS = 'diners'
DISCOVER = 'discover'
ELO = 'elo'
VISA = 'visa'

CARD_BRANDS = (MASTERCARD, DINERS, DISCOVER, ELO, VISA)


CREDITO_A_VISTA = '1'
PARCELADO_LOJA = '2'
PARCELADO_ADMINISTRADORA = '3'
DEBITO = 'A'

PRODUCTS = (
    CREDITO_A_VISTA,
    PARCELADO_LOJA,
    PARCELADO_ADMINISTRADORA,
    DEBITO,
)

NAO_INFORMADO = 'nao-informado'
ILEGIVEL = 'ilegivel'
INEXISTENTE = 'inexistente'
INFORMADO = 'informado'

ST_CRIADA = 0
ST_EM_ANDAMENTO = 1
ST_AUTENTICADA = 2
ST_NAO_AUTENTICADA = 3
ST_AUTORIZADA = 4
ST_NAO_AUTORIZADA = 5
ST_CAPTURADA = 6
ST_NAO_CAPTURADA = 8
ST_CANCELADA = 9
ST_EM_AUTENTICACAO = 10

STATUS = (
    ST_CRIADA,
    ST_EM_ANDAMENTO,
    ST_AUTENTICADA,
    ST_NAO_AUTENTICADA,
    ST_AUTORIZADA,
    ST_NAO_AUTORIZADA,
    ST_CAPTURADA,
    ST_NAO_CAPTURADA,
    ST_CANCELADA,
    ST_EM_AUTENTICACAO,
)


DEFAULT_CURRENCY = '096'


class Order(object):
    def __init__(self, value, number, datetime, description=None,
                 currency=DEFAULT_CURRENCY, language='PT'):
        self.value = value
        self.number = number
        self.datetime = datetime
        self.currency = currency
        self.description = description
        self.language = language


class Payment(object):
    def __init__(self, value, datetime, installments, card_brand, card_number,
                 card_holder_name, card_expiration_date, card_security_code):
        self.value = value
        self.datetime = datetime
        self.installments = installments
        self.card_brand = card_brand
        self.card_number = card_number
        self.card_holder_name = card_holder_name
        self.card_expiration_date = card_expiration_date
        self.card_security_code = card_security_code


class DebitPayment(Payment):
    def __init__(self, *args, **kwargs):
        kwargs.setdefault('installments', 1)
        super(Payment, self).__init__(*args, **kwargs)


class Response(object):
    @property
    def success(self):
        return not isinstance(self, Error)


class Transaction(Response):
    def __init__(self, order, status, authentication=None,
                 authorization=None, capture=None, cancel=None, pan=None):
        self.order = order
        #self.payment = payment
        self.status = status
        self.authentication = authentication
        self.authorization = authorization
        self.capture = capture
        self.cancel = cancel
        self.pan = pan

    @classmethod
    def deserialize(cls, appstruct):
        order = Order(**appstruct['order'])
        return cls(order=order, status=appstruct['status'])


class Error(Response):
    def __init__(self, code, message):
        self.code = code
        self.message = message

    @classmethod
    def deserialize(cls, appstruct):
        return cls(**appstruct)


class SecurityCodeIndicator(colander.Integer):
    map = dict((
        (NAO_INFORMADO, 0),
        (INFORMADO, 1),
        (ILEGIVEL, 2),
        (INEXISTENTE, 9),
    ))
    _rmap = dict((b, a) for (a, b) in map.iteritems())

    def serialize(self, node, appstruct):
        i = self.map.get(appstruct, None)
        if i is None:
            raise colander.Invalid(node, "%s is not a valid SecuirtyCodeIndicator" % appstruct)
        return super(SecurityCodeIndicator, self).serialize(node, i)

    def deserialize(self, node, cstruct):
        i = super(SecurityCodeIndicator, self).deserialize(node, cstruct)

        if i is colander.null:
            return i

        appstruct = self._rmap.get(i, None)

        if appstruct is None:
            raise colander.Invalid(node, "%s is not a valid SecurityCodeIndicator" % i)

        return appstruct


class Money(colander.Decimal):
    """
    Serializes python numeric values.

    Note that valid monetary values should not have more than two
    decimal places.

    ::

        >>> node = colander.SchemaNode(Money())
        >>> node.serialize(200)
        '20000'
        >>> node.serialize(200.0)
        '20000'
        >>> node.serialize(199.5)
        '19950'
        >>> node.serialize(200.21)
        '20021'
        >>> from decimal import Decimal
        >>> node.serialize(Decimal('200'))
        '20000'
        >>> node.serialize(Decimal('200.21'))
        '20021'
        >>> node.serialize(Decimal('200.3'))
        '20030'
        >>> node.serialize(200.213)
        Traceback (most recent call last):
          ...
        Invalid: {'': u'"200.213" is not a valid Money value because it has more than two decimal places'}
        >>> node.deserialize('20000')
        Decimal('200.00')
        >>> node.deserialize('193')
        Decimal('1.93')
        >>> node.deserialize('190')
        Decimal('1.90')
    """
    def serialize(self, node, appstruct):
        cstruct = super(Money, self).serialize(node, appstruct)

        if cstruct is not colander.null:
            if '.' in cstruct:
                i, d = str(cstruct).rsplit('.', 1)
            else:
                i = str(cstruct)
                d = '00'

            if len(d) > 2:
                raise colander.Invalid(node,
                    ('"%s" is not a valid Money value because it has'
                     ' more than two decimal places' % appstruct))
            elif len(d) < 2:
                d += '0' * (2 - len(d))

            return "%s%s" % (i, d)
        else:
            return cstruct

    def deserialize(self, node, cstruct):
        if cstruct:
            cstruct = "%s.%s" % (cstruct[:-2], cstruct[-2:])
        return super(Money, self).deserialize(node, cstruct)


class BaseDateTime(colander.SchemaType):
    err_template =  "Invalid date"
    format = None

    def serialize(self, node, appstruct):
        if not appstruct:
            return colander.null

        if type(appstruct) is datetime.date: # cant use isinstance; dt subs date
            appstruct = datetime.datetime.combine(appstruct, datetime.time())

        if not isinstance(appstruct, datetime.datetime):
            raise colander.Invalid(node, '"%s" is not a datetime object' % appstruct)

        return appstruct.strftime(self.format)

    def deserialize(self, node, cstruct):
        if not cstruct:
            return colander.null
        try:
            result = self._strptime(cstruct)
        except ValueError:
            raise colander.Invalid(node, self.err_template)

        return result

    def _strptime(self, cstruct):
        """
        Deve levantar ValueError caso a estrutura seja inválida.
        """
        return datetime.datetime.strptime(cstruct, self.format)


class DateTime(BaseDateTime):
    format = "%Y-%m-%dT%H:%M:%S"

    def _strptime(self, cstruct):
        # XXX nas respostas da Cielo, os timestamps contém milisegundos
        # e timezone. precisamos tratar isso.
        try:
            return super(DateTime, self)._strptime(cstruct)
        except ValueError:
            dt, extra = cstruct.split('.', 1)
            ts = super(DateTime, self)._strptime(dt)

            # FIXME vamos fazer nosso melhor para carregar o timestamp de
            # forma correta, mas vamos ignorar o timezone
            ms, tz = extra.split('-')
            ts += datetime.timedelta(milliseconds=int(ms))
            return ts


class Month(BaseDateTime):
    err_template =  "Invalid month"
    format = "%Y%m"
    extended_format = None


class CardHolderSchema(colander.Schema):
    """
    dados-portador

    dados-portador.numero           N   R   16      Número do cartão.
    dados-portador.validade         N   R   6       Validade do cartão no formato aaaamm.
                                                    Exemplos: 201212 (dez 2012).
    dados-portador.indicador        N   R   1       Indicador do código de segurança: 0 (não informado),
                                                    1 (informado), 2 (ilegível), 9 (inexistente).
    dados-portador.codigo-seguranca N   C   3..4    Obrigatório se indicador = 1.
    dados-portador.nome-portador    AN  O   0..50   Opcional. Nome impresso no cartão.
    """
    number = colander.SchemaNode(colander.String(),
                                 tag='numero',
                                 validator=colander.Length(16, 16))
    expiration_date = colander.SchemaNode(Month(), tag='validade')
    security_code_indicator = colander.SchemaNode(SecurityCodeIndicator(),
                                                  tag='indicador')
    security_code = colander.SchemaNode(colander.String(),
                                        tag='codigo-seguranca',
                                        validator=colander.Length(3, 4))
    holder_name = colander.SchemaNode(colander.String(),
                                      tag='nome-portador',
                                      missing=colander.null,
                                      validator=colander.Length(max=50))


class OrderSchema(colander.Schema):
    """
    dados-pedido

    dados-pedido.numero     AN  R   1..20   Número do pedido da loja. Recomenda-se que seja um valor único por pedido.
    dados-pedido.valor      N   R   1..12   Valor do pedido.
    dados-pedido.moeda      N   R   3       Código numérico da moeda na ISO 4217.
                                            Para o Real, o código é 986.
    dados-pedido.data-hora  AN  R   19      Data hora do pedido.
    dados-pedido.descricao  AN  O   0..1024 Descrição do pedido.
    dados-pedido.idioma     AN  O   2       Idioma do pedido: PT (português), EN (inglês) ou ES (espanhol).
                                            Com base nessa informação é definida a língua a ser utilizada
                                            nas telas da Cielo. Caso não preenchido, assume-se PT.
    """
    number = colander.SchemaNode(colander.String(),
                                 tag='numero',
                                 validator=colander.Length(max=20))
    value = colander.SchemaNode(Money(),
                                tag='valor',
                                validator=colander.Range(
                                    min=Decimal('0.01'),
                                    max=Decimal('9999999999.99')
                                ))
    currency = colander.SchemaNode(colander.String(), tag='moeda') # TODO OneOf
    datetime = colander.SchemaNode(DateTime(), tag='data-hora')
    description = colander.SchemaNode(colander.String(),
                                      tag='descricao',
                                      validator=colander.Length(max=1024),
                                      missing=colander.null)
    language = colander.SchemaNode(colander.String(),
                                   tag='idioma',
                                   missing='PT',
                                   defaults='PT',
                                   validator=colander.OneOf(LANGUAGES))


class PaymentSchema(colander.Schema):
    """
    forma-pagamento

    forma-pagamento.bandeira    AN  R   n/a     Bandeira: visa, mastercard, diners,
                                                discover ou elo (em minúsculo).
    forma-pagamento.produto     AN  R   1       Código do produto: 1 (Crédito à Vista),
                                                2 (Parcelado loja), 3 (Parcelado administradora),
                                                A (Débito).
    forma-pagamento.parcelas    N   R   1..3    Número de parcelas. Para crédito à vista ou
                                                débito, utilizar 1.
    """
    card_brand = colander.SchemaNode(colander.String(),
                                     tag='bandeira',
                                     validator=colander.OneOf(CARD_BRANDS))
    product = colander.SchemaNode(colander.String(),
                                  tag='produto',
                                  validator=colander.OneOf(PRODUCTS))
    installments = colander.SchemaNode(colander.Integer(),
                                       tag='parcelas',
                                       validator=colander.Range(1, 999))


class EstablishmentSchema(colander.Schema):
    """
    dados-ec

    dados-ec.numero N   R   1..20   Número de afiliação da loja com a Cielo.
    dados-ec.chave  AN  R   1..100  Chave de acesso da loja atribuída pela Cielo.
    """
    number = colander.SchemaNode(colander.String(),
                                 tag='numero',
                                 validator=colander.Length(max=20))
    key = colander.SchemaNode(colander.String(),
                              tag='chave',
                              validator=colander.Length(max=100))


class AutenticacaoSchema(colander.Schema):
    codigo = colander.SchemaNode(colander.Integer())
    mensagem = colander.SchemaNode(colander.String())
    data = colander.SchemaNode(DateTime(), tag='data-hora')
    valor = colander.SchemaNode(Money())
    eci = colander.SchemaNode(colander.Integer())


class AutorizacaoSchema(colander.Schema):
    codigo = colander.SchemaNode(colander.Integer())
    mensagem = colander.SchemaNode(colander.String())
    data = colander.SchemaNode(DateTime(), tag='data-hora')
    valor = colander.SchemaNode(Money())
    lr = colander.SchemaNode(colander.Integer()) # TODO OneOf?
    nsu = colander.SchemaNode(colander.String()) # TODO wtf?
    arp = colander.SchemaNode(colander.String(), missing=colander.null) # só está disponível em transações que foram auatorizadas


class CapturaSchema(colander.Schema):
    codigo = colander.SchemaNode(colander.Integer())
    mensagem = colander.SchemaNode(colander.String())
    data = colander.SchemaNode(DateTime(), tag='data-hora')
    valor = colander.SchemaNode(Money())


class CancelamentoSchema(colander.Schema):
    codigo = colander.SchemaNode(colander.Integer())
    mensagem = colander.SchemaNode(colander.String())
    data = colander.SchemaNode(DateTime(), tag='data-hora')
    valor = colander.SchemaNode(Money())


class Raiz(colander.Schema):
    """
    Schema base para schemas de nós-raiz.
    """
    id = colander.SchemaNode(colander.String(), attrib=True)
    version = colander.SchemaNode(colander.String(), tag='versao', attrib=True)


class TransactionRequestSchema(Raiz):
    establishment = EstablishmentSchema(tag='dados-ec')
    holder = CardHolderSchema(tag='dados-portador', missing=colander.null)
    order = OrderSchema(tag='dados-pedido')
    payment = PaymentSchema(tag='forma-pagamento')
    return_url = colander.SchemaNode(colander.String(), tag='url-retorno')
    authorize = colander.SchemaNode(colander.Integer(), tag='autorizar')
    capture = colander.SchemaNode(colander.Boolean(), tag='capturar', missing=colander.null)
    bin = colander.SchemaNode(colander.String(), tag='bin', missing=colander.null)


class RequisicaoConsultaSchema(Raiz):
    """
    Schema da Consulta via TID

    requisicao-consulta

    dados-ec.numero N   R   1..20   Número de afiliação da loja com a Cielo
    dados-ec.chave  AN  R   1..100  Chave de acesso da loja atribuída pela Cielo
    tid             AN  R   1..40   Identificador da transação.
    """
    tid = colander.SchemaNode(colander.String(), validator=colander.Length(max=40))
    ec = EstablishmentSchema(tag='dados-ec')


class RequisicaoConsultaOrderSchema(Raiz):
    """
    Schema da Consulta via Número do Pedido

    requisicao-consulta-chsec

    numero-pedido   NA  R   1...20  Número do Pedido associado à Transação
    dados-ec.numero N   R   1..20   Número de afiliação da loja com a Cielo
    dados-ec.chave  AN  R   1..100  Chave de acesso da loja atribuída pela Cielo
    """
    numero_pedido = colander.SchemaNode(colander.String(),
                                        tag='numero-pedido',
                                        validator=colander.Length(max=20))
    ec = EstablishmentSchema(tag='dados-ec')


class RequisicaoCapturaSchema(Raiz):
    """
    Schema da requisição de captura.

    requisicao-captura

    dados-ec.numero N   R   1..20   Número de afiliação da loja com a Cielo
    dados-ec.chave  AN  R   1..100  Chave de acesso da loja atribuída pela Cielo
    tid             AN  R   1..40   Identificador da transação.
    valor           N   O   1..12   Valor da captura. Caso não fornecido, o valor atribuído é o valor da autorização.
    anexo           AN  O   1..1024 Informação adicional para detalhamento da captura.
    """
    tid = colander.SchemaNode(colander.String(),
                              validator=colander.Length(max=40))
    valor = colander.SchemaNode(Money(),
                                missing=colander.null,
                                validator=colander.Range(max=Decimal('9999999999.99')))
    anexo = colander.SchemaNode(colander.String(),
                                missing=colander.null,
                                validator=colander.Length(max=1024))
    ec = EstablishmentSchema(tag='dados-ec')


class ErrorSchema(colander.Schema):
    tag = 'erro'
    content_type = Error

    code = colander.SchemaNode(colander.Integer(), tag='codigo')
    message = colander.SchemaNode(colander.String(), tag='mensagem')


class TransactionSchema(Raiz):
    tag = 'transacao'
    content_type = Transaction

    tid = colander.SchemaNode(colander.String())
    order = OrderSchema(tag='dados-pedido')
    payment = PaymentSchema(tag='forma-pagamento')
    status = colander.SchemaNode(colander.Integer(), validator=colander.OneOf(STATUS))
    authentication = AutenticacaoSchema(tag='autenticacao', missing=colander.null)
    authorization = AutorizacaoSchema(tag='autorizacao', missing=colander.null)
    capture = CapturaSchema(tag='captura', missing=colander.null)
    cancel = CancelamentoSchema(tag='cancelamento', missing=colander.null)
    pan = colander.SchemaNode(colander.String())
    authentication_url = colander.SchemaNode(colander.String(),
                                             tag='url-autenticacao',
                                             missing=colander.null)


def guess_response_schema(tag):
    from bbe.cielo import schema
    members = inspect.getmembers(schema)
    for (name, value) in members:
        if not isinstance(value, type):
            continue
        if not issubclass(value, colander.Schema):
            continue
        if getattr(value, 'tag', None) == tag:
            return value(tag=value.tag, content_type=value.content_type)
